File: backend/app/services/llm_service.py
import os
from typing import Dict
from dotenv import load_dotenv
import openai
import google.generativeai as genai  # Using stable generativeai library
from anthropic import Anthropic
from groq import Groq

load_dotenv()

class LLMService:
    def __init__(self):
        # Initialize clients if keys are present
        self.openai_key = os.getenv("OPENAI_API_KEY")
        if self.openai_key:
            self.openai_client = openai.OpenAI(api_key=self.openai_key)
            
        self.anthropic_key = os.getenv("ANTHROPIC_API_KEY")
        if self.anthropic_key:
            self.anthropic_client = Anthropic(api_key=self.anthropic_key)
            
        self.google_key = os.getenv("GOOGLE_API_KEY")
        if self.google_key:
            genai.configure(api_key=self.google_key)
            
        self.groq_key = os.getenv("GROQ_API_KEY")
        if self.groq_key:
            self.groq_client = Groq(api_key=self.groq_key)

    async def get_response(self, model_provider: str, model_name: str, prompt: str) -> str:
        """
        Fetches a response from the specified AI model.
        """
        # Map 'auto' to default model names
        default_models = {
            "openai": "gpt-4o",
            "google": "gemini-2.5-flash",  # Latest Gemini 2.5 Flash
            "anthropic": "claude-sonnet-4-5",  # Latest Claude Sonnet 4.5
            "deepseek": "deepseek-chat",
            "meta": "llama-3.1-70b-versatile"
        }
        
        if model_name == "auto" or not model_name:
            model_name = default_models.get(model_provider, "gpt-4o")
        
        try:
            if model_provider == "openai":
                if not self.openai_key: return "Error: OPENAI_API_KEY not configured"
                return await self._call_openai(model_name, prompt)
            elif model_provider == "google":
                if not self.google_key: return "Error: GOOGLE_API_KEY not configured"
                return await self._call_google(model_name, prompt)
            elif model_provider == "anthropic":
                if not self.anthropic_key: return "Error: ANTHROPIC_API_KEY not configured"
                return await self._call_anthropic(model_name, prompt)
            elif model_provider == "meta":
                if not self.groq_key: return "Error: GROQ_API_KEY (for Meta) not configured"
                return await self._call_meta(model_name, prompt)
            elif model_provider == "deepseek":
                # DeepSeek often uses OpenAI compatible API
                if not os.getenv("DEEPSEEK_API_KEY"): return "Error: DEEPSEEK_API_KEY not configured"
                return await self._call_deepseek(model_name, prompt)
            else:
                return f"Error: Unknown provider {model_provider}"
        except Exception as e:
            return f"Error calling {model_name}: {str(e)}"

    async def _call_openai(self, model: str, prompt: str) -> str:
        response = self.openai_client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}]
        )
        return response.choices[0].message.content

    async def _call_google(self, model_name: str, prompt: str) -> str:
        try:
            model = genai.GenerativeModel(model_name)
            response = model.generate_content(prompt)
            return response.text
        except Exception as e:
            return f"Error calling Google model {model_name}: {str(e)}"

    async def _call_anthropic(self, model_name: str, prompt: str) -> str:
        try:
            message = self.anthropic_client.messages.create(
                model=model_name,
                max_tokens=1024,
                messages=[
                    {
                        "role": "user",
                        "content": prompt
                    }
                ]
            )
            return message.content[0].text
        except Exception as e:
            return f"Error calling Anthropic model {model_name}: {str(e)}"

    async def _call_meta(self, model: str, prompt: str) -> str:
        # Using Groq for Meta Llama models
        chat_completion = self.groq_client.chat.completions.create(
            messages=[{"role": "user", "content": prompt}],
            model=model,
        )
        return chat_completion.choices[0].message.content

    async def _call_deepseek(self, model: str, prompt: str) -> str:
        client = openai.OpenAI(
            api_key=os.getenv("DEEPSEEK_API_KEY"), 
            base_url="https://api.deepseek.com"
        )
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}]
        )
        return response.choices[0].message.content

File: frontend/lib/services/llmService.ts
import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';
import { GoogleGenerativeAI } from '@google/generative-ai';

export class LLMService {
    private openai?: OpenAI;
    private anthropic?: Anthropic;
    private google?: GoogleGenerativeAI;

    constructor() { }

    private initOpenAI() {
        if (this.openai) return;
        if (process.env.OPENAI_API_KEY) {
            this.openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
        }
    }

    private initAnthropic() {
        if (this.anthropic) return;
        if (process.env.ANTHROPIC_API_KEY) {
            this.anthropic = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
        }
    }

    private initGoogle() {
        if (this.google) return;
        if (process.env.GOOGLE_API_KEY) {
            this.google = new GoogleGenerativeAI(process.env.GOOGLE_API_KEY);
        }
    }

    async getResponse(provider: string, modelName: string, prompt: string): Promise<string> {
        const defaultModels: Record<string, string> = {
            openai: 'gpt-4o',
            google: 'gemini-2.5-flash',
            anthropic: 'claude-sonnet-4-5',
            deepseek: 'deepseek-chat',
        };

        const model = modelName === 'auto' || !modelName ? defaultModels[provider] || 'gpt-4o' : modelName;

        try {
            switch (provider) {
                case 'openai':
                    return await this.callOpenAI(model, prompt);
                case 'google':
                    return await this.callGoogle(model, prompt);
                case 'anthropic':
                    return await this.callAnthropic(model, prompt);
                case 'deepseek':
                    return await this.callDeepseek(model, prompt);
                default:
                    return `Error: Unknown provider ${provider}`;
            }
        } catch (error: any) {
            console.error(`Error calling ${model}:`, error);
            return `Error calling ${model}: ${error.message}`;
        }
    }

    private async callOpenAI(model: string, prompt: string): Promise<string> {
        this.initOpenAI();
        if (!this.openai) throw new Error('OpenAI API key not configured');
        const response = await this.openai.chat.completions.create({
            model,
            messages: [{ role: 'user', content: prompt }],
        });
        return response.choices[0].message.content || '';
    }

    private async callGoogle(modelName: string, prompt: string): Promise<string> {
        this.initGoogle();
        if (!this.google) throw new Error('Google API key not configured');
        const model = this.google.getGenerativeModel({ model: modelName });
        const result = await model.generateContent(prompt);
        return result.response.text();
    }

    private async callAnthropic(modelName: string, prompt: string): Promise<string> {
        this.initAnthropic();
        if (!this.anthropic) throw new Error('Anthropic API key not configured');
        const message = await this.anthropic.messages.create({
            model: modelName,
            max_tokens: 1024,
            messages: [{ role: 'user', content: prompt }],
        });

        const content = message.content[0];
        if (content.type === 'text') {
            return content.text;
        }
        return '';
    }

    private async callDeepseek(model: string, prompt: string): Promise<string> {
        const deepseek = new OpenAI({
            apiKey: process.env.DEEPSEEK_API_KEY,
            baseURL: 'https://api.deepseek.com',
        });
        const response = await deepseek.chat.completions.create({
            model,
            messages: [{ role: 'user', content: prompt }],
        });
        return response.choices[0].message.content || '';
    }
}

export const llmService = new LLMService();
